### PR TITLE
Reduce input buffer size

### PR DESCRIPTION
**Pull request by waywardgeek**
This change was suggested by Kurt Roeckx during his code review.  The double-sized buffer bothered me before, but I did not see how easy it is to reduce it by 2X.  Unless I've made a mistake (warning: I make a lot!) then this patch not only simplifies the code slightly, but it also now runs slightly faster, and saves a few bytes of memory.

### CODE REVIEW

File: ref/blake2b-ref.c [hunk 1]
```diff
@@ -327,7 +326,7 @@ int blake2b_final( blake2b_state *S, uint8_t *out, uint8_t outlen )
```
Comment (kroeckx) [inline ref/blake2b-ref.c]: The if above should also be removed, like you did in the other cases.

Comment (waywardgeek) [inline ref/blake2b-ref.c]: Good catch.  Thanks for the careful review!

**Comment by sneves** [comment]
Two objections:

 - Consider the following sequence:
```
blake2b_init(S, ...);
blake2b_update(S, some_buffer, 128);
blake2b_final(S, hash, 64);
```
With your patch, we will get a compression call during `blake2b_update`, seeing that a new complete block is input, and we'll also have a compression call during `blake2b_final`, rendering the result incorrect---only one compression should happen for a 128-byte input.

 - Perhaps more important to performance than avoiding the double buffer, which I agree should go away, is to avoid unnecessary copies to buffer when compressing long inputs. I've already solved this (and the above issue) [in this branch of libb2](https://github.com/BLAKE2/libb2/blob/17f2a5ebc97793c62e93146095dc5e34aa45a206/src/portable/blake2b.c#L234-L253).

**Comment by sneves** [comment]
Actually, my first issue above is incorrect; there is a single compression in that case.

**Comment by waywardgeek** [comment]
On Mon, Mar 7, 2016 at 4:47 PM, Samuel Neves <notifications@github.com>
wrote:

> Two objections:
>
>    - Consider the following sequence:
>
> blake2b_init(S, ...);
> blake2b_update(S, some_buffer, 128);
> blake2b_final(S, hash, 64);
>
> With your patch, we will get a compression call during blake2b_update,
> seeing that a new complete block is input, and we'll also have a
> compression call during blake2b_final, rendering the result
> incorrect---only one compression should happen for a 128-byte input.
>
>    - Perhaps more important to performance than avoiding the double
>    buffer, which I agree should go away, is to avoid unnecessary copies to
>    buffer when compressing long inputs. I've already solved this (and the
>    above issue) in this branch of libb2
>    <https://github.com/BLAKE2/libb2/blob/17f2a5ebc97793c62e93146095dc5e34aa45a206/src/portable/blake2b.c#L234-L253>
>    .
>
> The version in your branch looks good to me.  That's fortunate, because I
was just told not to offer patches to BLAKE2, because the tripple-license
includes cc0 :-/  I'll close the PR.  Man, lawyers make it _hard_ to
volunteer to do good for the world.

**Comment by zooko** [comment]
Whoa whoa whoa -- who told you not to offer patches to BLAKE2? Fuck that
noise. Offer patches to BLAKE2!

**Comment by waywardgeek** [comment]
On Mon, Mar 7, 2016 at 6:15 PM, zooko <notifications@github.com> wrote:

> Whoa whoa whoa -- who told you not to offer patches to BLAKE2? Fuck that
> noise. Offer patches to BLAKE2!
>

It was just 50-ish lines, and Samuel did an even nicer version already in
his libb2 branch.  It's just big company nonsense.  Still, these lawyers
are a joy-kill to collaborating to improve the world.

Bill

**Pull request closed without merging**